### DISCUSSION TITLE
The last ISAKMP packet and the first ESP packet arrive at the device with a time interval that is too brief. The strongSwan device is unable to fully process the ISAKMP packet before the ESP packet arrives. Is there a way to prevent this situation from occurring?

### DISCUSSION THREAD

**Discussion by YuvalCheung** [strongswan/strongswan#2514] [Q&A] [answered]
When I use the strongSwan 5.7.2 device (10.10.10.1) to establish a connection with a test instrument, the test instrument will establish a link with the strongSwan device. The time at which the last ISAKMP packet arrives at the strongSwan device is 30.626389, and the arrival time of the first ESP packet is 30.626567. The time difference between these two packets arriving at the strongSwan device is less than 1 millisecond, which causes the strongSwan device to be unable to process the ESP packet normally. As a result, 3 seconds later, the test instrument on the other side performs a link disconnection operation.
Is there a solution in strongSwan for this situation, such as "buffering" the ESP packets?

![202410290914400252170280170893FD](https://github.com/user-attachments/assets/62600c14-ff5f-4f3e-aad9-e11e97b64efb)

And upon checking `cat /proc/1/net/xfrm_stat`, the value of XfrmAcquireError increases (it shows 17 because I have tried multiple times).

![202410282336160252170260166253FD](https://github.com/user-attachments/assets/47796dc6-ca0d-4a4d-a1d1-b05e002c55b7)

This phenomenon is very similar to the issue described in https://github.com/strongswan/strongswan/discussions/2467.

**Comment by tobiasbrunner** [comment, marked as answer]
> The time difference between these two packets arriving at the strongSwan device is less than 1 millisecond, which causes the strongSwan device to be unable to process the ESP packet normally.

That can definitely happen but is not usually a problem.

> As a result, 3 seconds later, the test instrument on the other side performs a link disconnection operation

That sounds like a mistake in that tool. IP is an unreliable protocol, so is IPsec. Any packet could get dropped. So why doesn't this tool send any retransmits within those three seconds?

> Is there a solution in strongSwan for this situation, such as "buffering" the ESP packets?

No (and it would have to be done in the Linux kernel anyway).

> And upon checking `cat /proc/1/net/xfrm_stat`, the value of XfrmAcquireError increases (it shows 17 because I have tried multiple times).

For the inbound SA, only an SPI has been allocated yet, so it is in state `XFRM_STATE_ACQ`, which causes this error when packets with a matching SPI arrive (until the SA is updated with the algorithm and keys etc. and is switched to state `XFRM_STATE_VALID`).

**Reply by YuvalCheung** [reply]
Because the ESP packet encapsulates a UDP packet, it will not be retransmitted. I encountered this issue in a performance testing scenario (it would not happen in other normal scenarios). After investigation, I suspect that when the ESP packet arrives, the tunnel establishment is not yet complete. I checked the related log prints of strongSwan and found that the IP xfrm policy needs to be configured after 9 packet exchanges. During the configuration period, if ESP packets arrive, they might be processed abnormally. However, since this ESP packet is a UDP packet, the testing instrument does not perform any retransmission action and directly determines the test as failed.

Therefore, I'm wondering if it's possible to implement a packet caching mechanism to complete this test scenario. If strongSwan does not support this kind of caching, I may need to establish a packet caching mechanism based on SPI in the kernel myself. Thank you for your answer.

**Reply by tobiasbrunner** [reply]
> Because the ESP packet encapsulates a UDP packet, it will not be retransmitted. ... However, since this ESP packet is a UDP packet, the testing instrument does not perform any retransmission action and directly determines the test as failed.

That seems like a flaw in that testing tool. UDP and IP/IPsec are unreliable protocols, so packets could get dropped at any point. Retransmits or repeated test packets are necessary in such protocols (just look at IKE, which is also UDP based, or some heartbeat-like protocol). If you can't change that, maybe delay the start of the test? Or initiate the test from the initiator of the SA?

**Reply by YuvalCheung** [reply]
I also want to conduct the test according to your suggestion, but unfortunately, the logic of this testing tool is not under my control. I can only choose to adapt. In any case, thank you for your answer.
